### PR TITLE
Fix outstanding futures clear bottleneck

### PR DESCRIPTION
**Pull request by pankajroark**
When maxWaitingFutures is high, AsyncBase can end up spending a lot of compute just clearing finished futures. The cost of clearing the finished futures is proportional to total number of outstanding futures.  If the number of finished futures is very low compared to total outstanding then we can end up paying this price for nothing much. And the same pattern can repeat on every execute. I found this happening in one of the jobs. In this change we keep track of number of pending futures and use that info to make sure that we clear the finished futures only when they are a good proportion of total number of outstanding futures so that we get good return on investment.

### CODE REVIEW

File: summingbird-online/src/main/scala/com/twitter/summingbird/online/executor/AsyncBase.scala [hunk 1]
```diff
@@ -80,20 +94,32 @@ abstract class AsyncBase[I, O, S, D, RC](maxWaitingFutures: MaxWaitingFutures, m
   private def addOutstandingFuture(fut: Future[Unit]): Boolean =
     if (!fut.isDefined) {
       outstandingFutures.put(fut)
+      numPendingOutstandingFutures.addAndGet(1)
+      fut.respond { _ => numPendingOutstandingFutures.addAndGet(-1) }
```
Comment (johnynek) [inline summingbird-online/src/main/scala/com/twitter/summingbird/online/executor/AsyncBase.scala]: can we use `.ensure(numPendingOutstandingFutures.decrementAndGet())`?

File: summingbird-online/src/main/scala/com/twitter/summingbird/online/executor/AsyncBase.scala [hunk 2]
```diff
@@ -80,20 +94,32 @@ abstract class AsyncBase[I, O, S, D, RC](maxWaitingFutures: MaxWaitingFutures, m
   private def addOutstandingFuture(fut: Future[Unit]): Boolean =
     if (!fut.isDefined) {
       outstandingFutures.put(fut)
+      numPendingOutstandingFutures.addAndGet(1)
```
Comment (johnynek) [inline summingbird-online/src/main/scala/com/twitter/summingbird/online/executor/AsyncBase.scala]: can we use `numPendingOutstandingFutures.incrementAndGet()`

File: summingbird-online/src/main/scala/com/twitter/summingbird/online/executor/AsyncBase.scala [hunk 3]
```diff
@@ -16,13 +16,26 @@ limitations under the License.
...
 
+object AsyncBase {
+  /**
+   * Ratio of total number of outstanding futures to the portion that is finished
+   * ,at which finished futures are cleared.
+   * Clearing finished futures costs proportional to total number of outstanding
+   * futures, so we want to make sure we only clear when sufficient portion is
+   * finished.
+   */
+  val OutstandingFuturesDequeueRatio = 2
```
Comment (johnynek) [inline summingbird-online/src/main/scala/com/twitter/summingbird/online/executor/AsyncBase.scala:36]: should this be a tunable option with a default of 2?

**Comment by johnynek** [comment]
couple of small comments.

File: summingbird-online/src/main/scala/com/twitter/summingbird/online/executor/AsyncBase.scala (hunk 3 above)
Comment (johnynek) [inline summingbird-online/src/main/scala/com/twitter/summingbird/online/executor/AsyncBase.scala:36]: maybe not for now. We can revisit if this does not fix the issue. Less knobs are better.

File: summingbird-online/src/main/scala/com/twitter/summingbird/online/executor/AsyncBase.scala (hunk 1 above)
Comment (pankajroark) [inline summingbird-online/src/main/scala/com/twitter/summingbird/online/executor/AsyncBase.scala]: Good idea.

File: summingbird-online/src/main/scala/com/twitter/summingbird/online/executor/AsyncBase.scala (hunk 3 above)
Comment (pankajroark) [inline summingbird-online/src/main/scala/com/twitter/summingbird/online/executor/AsyncBase.scala:36]: I agree, making this a knob seemed overkill to me as well.

**Comment by pankajroark** [comment]
@johnynek I incorporated comments, could you take a look again? Thanks

**Pull request merged**